base/set.jl

```diff
@@ -167,15 +167,39 @@ julia> unique(x -> x^2, [1, -1, 3, -3, 4])
 """
 function unique(f, C)
     out = Vector{eltype(C)}()
-    seen = Set()
-    for x in C
+
+    s = iterate(C)
+    if s === nothing
+        return out
+    end
+    (x, i) = s
+    y = f(x)
+    seen = Set{typeof(y)}()
+    push!(seen, y)
+    push!(out, x)
+
+    return _unique!(f, out, C, seen, i)
+end
+
+function _unique!(f, out::AbstractVector, C, seen::Set, i)
+    s = iterate(C, i)
+    while s !== nothing
+        (x, i) = s
         y = f(x)
-        if !in(y, seen)
-            push!(seen, y)
+        if y ∉ seen
             push!(out, x)
+            if y isa eltype(seen)
+                push!(seen, y)
+            else
```

Comment:
Can we do 
```
else
    TT = promote_typejoin(eltype(seen), typeof(y))
    if y isa TT
        push!(seen, y)
        else
             seen2 = ...
        end
end
```
Idea is that we don't grow the stack if `f` is e.g. type-unstable between `Float32` and `Float64`.

Reply:
Eh, maybe disregard. I misunderstoof `promote_typejoin`, which apparently doesn't do that. Point remains, do we want collect-style widening here (integer and float merge to `Float64`) or do we want exact widening as you did?

Reply:
Wait, is this calling `_unique!` recursively each time a new value is encountered?! @chethega is right that this is bad, a call should only happen when encountering a new *type*.

Reply:
No - it simply pushes if the value is of type eltype of seen. If not, and only if not, then it constructs seen2 and makes a new call.